src/lib/battery/battery.cpp

```diff
@@ -188,35 +188,128 @@ void Battery::sumDischarged(const hrt_abstime &timestamp, float current_a)
 
 void Battery::estimateStateOfCharge(const float voltage_v, const float current_a, const float throttle)
 {
-	// remaining battery capacity based on voltage
-	float cell_voltage = voltage_v / _params.n_cells;
 
-	// correct battery voltage locally for load drop to avoid estimation fluctuations
-	if (_params.r_internal >= 0.f) {
-		cell_voltage += _params.r_internal * current_a;
-
-	} else {
-		// assume linear relation between throttle and voltage drop
-		cell_voltage += throttle * _params.v_load_drop;
+	bool sees_coulomb_counting_only = true;
+	// Sees.ai modification coulomb_counting_only
+	// Set to 'true' for Sees behaviour:
+	// SoC relies solely on mAh discharged.
+	//
+	// Set to 'false' for default PX4 behaviour:
+	// SoC is based on whichever of the two options below that provides the lower value:
+	// - A combination of discharge and Voltage based estimation (relies more on voltage at lower voltages).
+	// - Just on discharge.
+
+
+	// -------------------------------------------
+	// If bool is true, do sees desired behaviour
+	// -------------------------------------------
+
+	if (sees_coulomb_counting_only) {
+
+		// This ensures the cell voltage is derived from the filtered voltage and corrects for any voltage load drop as PX4 uses the raw voltage values.
+
+		cell_voltage_filtered = _voltage_filter_v.getState();
+		cell_voltage_filtered /= _params.n_cells;
+		cell_voltage_filtered_load += _params.r_internal * current_a;
+
+		// Derives initial SoC Estimation from the voltage. A 15 second timer is included to allow voltage to settle whilst components initialise.
+		if(first_run || ((hrt_absolute_time() - first_run_time) < 15'000'000)){
+			struct Lookup {
+				float OCVoltage;
+				float SOC;
+			};
+				Lookup SOCLookup[] {
+				{4.17, 100.0},
+				{4.09, 89.5},
+				{3.99, 79.1},
+				{3.93, 68.6},
+				{3.87, 58.2},
+				{3.82, 47.7},
+				{3.79, 37.3},
+				{3.77, 26.8},
+				{3.73, 16.4},
+				{3.69, 5.9},
+				{3.50, 0.0}
+			};
+			bool above_min = true;
+			bool below_max = true;
+
+
+
+			if (cell_voltage_filtered >= SOCLookup[0].OCVoltage) {
+				_state_of_charge = 1.0;
+				below_max = false;
+			}
+			if (cell_voltage_filtered <= SOCLookup[10].OCVoltage) {
+				_state_of_charge = 0.0;
+				above_min = false;
+			}
+			if (above_min && below_max) {
+				for (int i = 0; i < 10; i++) {
+					float voltage_key = SOCLookup[i].OCVoltage;
+					if (cell_voltage_filtered > voltage_key) {
+						const float volt1 = SOCLookup[i].OCVoltage;
+						const float soc1 = SOCLookup[i].SOC;
+						const float volt2 = SOCLookup[i+1].OCVoltage;
+						const float soc2 = SOCLookup[i+1].SOC;
+						soc_initial = (soc1 + (soc2 - soc1) * (cell_voltage_filtered - volt1) / (volt2 - volt1))/100;
+						break;
+					}
+				}
+			}
+			if (first_run) {
```

Comment:
Small point - but this should be before the `if` statement above, otherwise `first_run_time` is undefined on first loop (will likely be a completely random value as it's not defined in the header, which isn't a big issue in this case, but good to avoid).